### PR TITLE
keyring HTTP API

### PR DESCRIPTION
**Pull request by tgross**
Implements the secure variables keyring HTTP API.

### CODE REVIEW

File: command/agent/keyring_endpoint.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// KeyringRequest is used route operator/raft API requests to the implementing
+// functions.
+func (s *HTTPServer) KeyringRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
```
Comment (tgross) [inline command/agent/keyring_endpoint.go:15]: Note that the RFC was missing (until just now) the URL paths for this HTTP API, so I'm totally open to changing those.

File: command/agent/keyring_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+	path := strings.TrimPrefix(req.URL.Path, "/v1/operator/keyring/")
+	switch {
+	case strings.HasPrefix(path, "keys"):
+		switch req.Method {
+		case http.MethodGet:
+			return s.keyringListRequest(resp, req)
+		case http.MethodPost, http.MethodPut:
```
Comment (angrycub) [inline command/agent/keyring_endpoint.go:23]: My only concern about doubling the ways to access the same endpoint is that it creates another API route that doesn't provide any end-user value, could actually create confusion, and prevents us from using the verb later should we want to.

File: command/agent/keyring_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+			return s.keyringUpsertRequest(resp, req, "")
+		default:
+			return nil, CodedError(405, ErrInvalidMethod)
+		}
+	case strings.HasPrefix(path, "key"):
+		keyID := strings.TrimPrefix(req.URL.Path, "/v1/operator/keyring/key/")
+		switch req.Method {
+		case http.MethodPost, http.MethodPut:
```
Comment (angrycub) [inline command/agent/keyring_endpoint.go:31]: See above comment re:route doubling

**Review by angrycub — APPROVED** [review]
Everything looked good to me.  Just had the one piece of feedback about tightening up our API routing in general.

File: command/agent/keyring_endpoint.go (hunk 2 above)
Comment (tgross) [reply command/agent/keyring_endpoint.go:23]: Yeah I felt conflicted about that. I'll remove the `/v1/operator/keyring/key/` version.

**Pull request merged**